File: src/Core.TypeScript/watermark/watermark.ts
// Watermark — the event-time watermark of Akidau et al. (The Dataflow Model, VLDB 2015), TypeScript oracle.
// Conforms to the F# canonical shape (src/Core/Watermark.fs) by agreeing on the shared seed
// (./golden-vectors.json) that the C#/F#/Rust oracles also verify. All integer arithmetic — no floats,
// byte-lockable in the safe-integer range (the .NET/Rust oracles use int64).

export type Strategy = "monotonic" | "bounded";

/**
 * The WatermarkTracker fold: returns the emitted watermark after each observed event time.
 * maxSeen = running max; candidate = maxSeen (monotonic) or maxSeen - lateness (bounded); clamped
 * monotone non-decreasing. The seed stays within the safe-integer range, so the MinValue sentinel
 * never surfaces in outputs.
 */
export function observe(strategy: Strategy, lateness: number, events: number[]): number[] {
  let maxSeen = Number.NEGATIVE_INFINITY;
  let lastEmitted = Number.NEGATIVE_INFINITY;
  const out: number[] = [];
  for (const e of events) {
    if (e > maxSeen) maxSeen = e;
    const candidate = strategy === "monotonic" ? maxSeen : maxSeen - lateness;
    if (candidate > lastEmitted) lastEmitted = candidate;
    out.push(lastEmitted);
  }
  return out;
}

/** Is eventTime late according to the current watermark? */
export function isLate(wm: number, eventTime: number): boolean {
  return eventTime <= wm;
}

/** Combine per-source watermarks downstream: min (can't progress past the slowest input). */
export function combine(sources: number[]): number {
  let min = Number.POSITIVE_INFINITY;
  let any = false;
  for (const s of sources) {
    any = true;
    if (s < min) min = s;
  }
  return any ? min : Number.NEGATIVE_INFINITY;
}

File: src/Core.CSharp/Watermark.cs
using System.Collections.Generic;

namespace Zeta.Core.CSharp;

/// <summary>
/// Watermark — the event-time watermark of Akidau et al. (The Dataflow Model, VLDB 2015), C# oracle.
/// Conforms to the F# canonical shape (<c>src/Core/Watermark.fs</c>) by agreeing on the shared seed
/// (<c>src/Core.TypeScript/watermark/golden-vectors.json</c>) that the F#/TS/Rust oracles also verify.
/// All <c>long</c> arithmetic — no floats, byte-lockable in the safe-integer range.
/// </summary>
public static class Watermark
{
    /// <summary>
    /// The <c>WatermarkTracker</c> fold: the emitted watermark after each observed event time.
    /// maxSeen = running max; candidate = maxSeen (monotonic) or maxSeen - lateness (bounded; the
    /// Periodic formula too); clamped monotone non-decreasing.
    /// </summary>
    public static IReadOnlyList<long> Observe(string strategy, long lateness, IReadOnlyList<long> events)
    {
        var maxSeen = long.MinValue;
        var lastEmitted = long.MinValue;
        var outp = new List<long>(events.Count);
        foreach (var e in events)
        {
            if (e > maxSeen)
            {
                maxSeen = e;
            }

            var candidate = string.Equals(strategy, "monotonic", System.StringComparison.Ordinal)
                ? maxSeen
                : (maxSeen == long.MinValue ? long.MinValue : maxSeen - lateness);
            if (candidate > lastEmitted)
            {
                lastEmitted = candidate;
            }

            outp.Add(lastEmitted);
        }

        return outp;
    }

    /// <summary>Is <paramref name="eventTime"/> late according to the current watermark?</summary>
    public static bool IsLate(long wm, long eventTime) => eventTime <= wm;

    /// <summary>Combine per-source watermarks downstream: min (can't progress past the slowest input).</summary>
    public static long Combine(IReadOnlyList<long> sources)
    {
        var min = long.MaxValue;
        var any = false;
        foreach (var s in sources)
        {
            any = true;
            if (s < min)
            {
                min = s;
            }
        }

        return any ? min : long.MinValue;
    }
}
